File: pde/proj2/Python_Code/thermal_fin.py
import numpy as np
from scipy.sparse import lil_matrix
from scipy.sparse.linalg import spsolve as solve


def thermal_fin(mesh, mu):

    """
    -------------------------------------------------------------------------

    ThermalFin.py

    -------------------------------------------------------------------------
    Computes the temperature distribution and root temperature for a fin
    using the Finite Element Method.
    -------------------------------------------------------------------------

    INPUT   mesh    grid label (coarse, medium, or fine)
            mu      thermal conductivities of sections and Biot number 1x5

    OUTPUT  u       temperature discretization in the fin
            Troot   root temperature

    -------------------------------------------------------------------------
    """

    # Parameters setup: rearrange the values in mu
    # kappa = [k1, k2, k3, k4, k0, Bi]

    kappa = np.ones(6)
    kappa[:4] = mu[:4]
    kappa[5] = mu[4]       # Bi

    A = lil_matrix((mesh['n_nodes'], mesh['n_nodes']))
    F = lil_matrix((mesh['n_nodes'], 1))

    # Domain Interior
    for i in range(5):
        for n in range(len(mesh['theta'][i])):
            phi = mesh['theta'][i][n,:]
            """
            Fill me in!
            """
            A[phi[:, None], phi] += A_local

    # Boundaries (not root)
    i=5
    for n in range(len(mesh['theta'][i])):
        phi = mesh['theta'][i][n,:]
        """
        Fill me in!
        """
        A[phi[:, None], phi] += A_local

    # Root boundary
    i=6
    for n in range(len(mesh['theta'][i])):
        phi = mesh['theta'][i][n,:]
        """
        Fill me in!
        """
        F[phi] = F[phi] + F_local[:, None]


    # Solve for the temperature distribution
    u = solve(A.asformat('csc'), F.asformat('csc'))

    # Compute the temperature at the root
    # Fill me in!

    return Troot, u

File: pde/proj2/Matlab Code/ThermalFin.m
function [u, Troot] = ThermalFin(mesh, mu)
%
% -------------------------------------------------------------------------
%   Computes the temperature distribution and root temperature for a fin
%   using the Finite Element Method.
% -------------------------------------------------------------------------
%
%   INPUT   mesh    grid lebel (coarse, medium, or fine)
%           mu      thermal conductivities of sections and Biot number 1x5
%
%   OUTPUT  u       temperature disctribution in the fin
%           Troot   root temperature
%
% -------------------------------------------------------------------------


% Parameters setup: rearranges the values in mu
kappa = ones(6,1);
kappa(1:4) = mu(1:4);    % Bi = mu(6)
kappa(6) = mu(5);


% Initialization
A = sparse(mesh.nodes, mesh.nodes);
F = sparse(mesh.nodes,1);


% Domain Interior
for i = 1:5         % interior regions
    for n = 1:length(mesh.theta{i})
        % Get mesh node index phi
        phi = mesh.theta{i}(n,:)';
        % Get the node coordinates for the current element
        vert = mesh.coor(phi, :);
        % Calculate the area of the triangle
        area = polyarea(vert(:,1), vert(:,2));
        % Shape functions derivatives
        G = [vert(2,2)-vert(3,2), vert(3,2)-vert(1,2), vert(1,2)-vert(2,2);
         vert(3,1)-vert(2,1), vert(1,1)-vert(3,1), vert(2,1)-vert(1,1)] / (2*area);    
        % Compute the element stiffness matrix
        Alocal = area * (G' * G);
        % update
        A(phi,phi) = A(phi,phi) + mu(i)*Alocal;    
    end
end


% Boundaries (not root)
i = 6;
    for n = 1:length(mesh.theta{i})
        phi = mesh.theta{i}(n,:)';
        % Get the node coordinates for the current element
        vert = mesh.coor(phi, :);
        % Distance
        dist = sqrt(sum((vert(1,:) - vert(2,:)).^2));
        Alocal = [1/3,1/6;1/6,1/3]*dist;
        % update
        A(phi,phi) = A(phi,phi) + mu(5)*Alocal;
    end


% Root Boundary
i = 7;
    for n = 1:length(mesh.theta{i})
        phi = mesh.theta{i}(n,:)';
        % Get the node coordinates for the current element
        vert = mesh.coor(phi, :);
        % Distance
        dist = sqrt(sum((vert(1,:) - vert(2,:)).^2));
        Flocal = [1/2;1/2]*dist;
        % update
        F(phi) = F(phi) + Flocal;
    end


% Solve for the temperature distribution
u = full(A\F);      % use full since A and F are sparse


% Compute the temperature at the root
Troot = F'*u;
    